### PR TITLE
Implement unknown6 and signature encryption

### PR DESCRIPTION
**Pull request by Gabrola**
I've implemented the signature encryption algorithm in c# so there is no need to p/invoke the C encrypt.dll
I've confirmed that this works on my bot and everything works perfectly.

### CODE REVIEW

File: PokemonGo.RocketAPI/POGOProtos/Networking/Envelopes/Signature.cs [hunk 1]
```diff
@@ -0,0 +1,2580 @@
...
+    /// <summary>File descriptor for POGOProtos/Networking/Envelopes/Signature.proto</summary>
+    public static pbr::FileDescriptor Descriptor {
+      get { return descriptor; }
+    }
+    private static pbr::FileDescriptor descriptor;
+
+    static SignatureReflection() {
+      byte[] descriptorData = global::System.Convert.FromBase64String(
+          string.Concat(
+            "Ci9QT0dPUHJvdG9zL05ldHdvcmtpbmcvRW52ZWxvcGVzL1NpZ25hdHVyZS5w",
```
Comment (wertzui) [inline PokemonGo.RocketAPI/POGOProtos/Networking/Envelopes/Signature.cs:25]: Why are you pasting the string as base64 here?

Comment (Gabrola) [inline PokemonGo.RocketAPI/POGOProtos/Networking/Envelopes/Signature.cs:25]: This whole file is compiled, I didn't paste anything in there. If you're curious, this is a protofile descriptor which is needed when you are working with message types not known at compile-time or you want to access fields at runtime using reflection. Either way, this is not needed in this application since we are already using the compiled classes, it's only there because the protobuf compiler did.

**Comment by Panthere** [comment]
Wow, thats quite an amazing port from  the C++ "_"

Good job.
I think I like this one a bit less than the other, but if the crypt.cs was included in the other I think it would work quite good. Perhaps a class for building the u6 sig?

File: PokemonGo.RocketAPI/Helpers/Utils.cs [hunk 2]
```diff
@@ -13,32 +13,36 @@ public static ulong FloatAsUlong(double value)
...
         {
-            uint firstHash = xxHash.CalculateHash(Encoding.ASCII.GetBytes(authTicket), -1, 0x1B845238);
+            var seed = BitConverter.ToUInt32(new xxHash(32, 0x1B845238).ComputeHash(authTicket), 0);
+            var xxh32 = new xxHash(32, seed);
 
             var locationBytes = new List<byte>();
-            locationBytes.AddRange(BitConverter.GetBytes(lat));
-            locationBytes.AddRange(BitConverter.GetBytes(lng));
-            locationBytes.AddRange(BitConverter.GetBytes(alt));
+            locationBytes.AddRange(BitConverter.GetBytes(lat).Reverse());
```
Comment (ArcticXWolf) [inline PokemonGo.RocketAPI/Helpers/Utils.cs]: Why did you change this? Why is this array reversing necessery? :)

File: PokemonGo.RocketAPI/Helpers/Utils.cs [hunk 3]
```diff
@@ -9,5 +12,37 @@ public static ulong FloatAsUlong(double value)
...
+
+        public static uint GenerateLocation1(byte[] authTicket, double lat, double lng, double alt)
+        {
+            var seed = BitConverter.ToUInt32(new xxHash(32, 0x1B845238).ComputeHash(authTicket), 0);
+            var xxh32 = new xxHash(32, seed);
+
+            var locationBytes = new List<byte>();
+            locationBytes.AddRange(BitConverter.GetBytes(lat).Reverse());
+            locationBytes.AddRange(BitConverter.GetBytes(lng).Reverse());
+            locationBytes.AddRange(BitConverter.GetBytes(alt).Reverse());
```
Comment (ArcticXWolf) [inline PokemonGo.RocketAPI/Helpers/Utils.cs:24]: Why is the reversing of the byte array necessary? :)

Comment (Gabrola) [inline PokemonGo.RocketAPI/Helpers/Utils.cs:24]: Values are stored in little endian format in memory on most computer architectures, including x86, which is reverse order from a human perspective. For some reason Niantic decided to use big-endian order for their hashing function so we have to comply and reverse the normal order. So a value of 0x01020304 is stored in memory as 04 03 02 01 so we have to reverse it. In the python implementation they're converting the value to a human-readable hex string and then split that up into a byte array.

Comment (ArcticXWolf) [inline PokemonGo.RocketAPI/Helpers/Utils.cs:24]: Ah, okay I though GetBytes automatically converts that back, but so this makes sense. Thank you :)
